src/bert_layers/attention.py

```diff
@@ -404,7 +408,226 @@ def forward(
+            raise ValueError(
+                f"The hidden size ({config.hidden_size}) is not a multiple of the number of attention "
+                f"heads ({config.num_attention_heads})"
+            )
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attn_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attn_head_size
+        self.p_dropout = config.attention_probs_dropout_prob
+        self.Wqkv = nn.Linear(config.hidden_size, 3 * self.all_head_size)
+        self.Wo = nn.Linear(config.hidden_size, config.hidden_size)
+        self.out_drop = (
+            nn.Dropout(config.attn_out_dropout_prob) if config.attn_out_dropout_prob > 0.0 else nn.Identity()
+        )
+
+        if config.rotary_emb_dim is None:
+            config.rotary_emb_dim = self.attn_head_size
+
+        assert RotaryEmbedding is not None, "rotary_emb is not installed"
+        self.rotary_emb = RotaryEmbedding(
+            config.rotary_emb_dim,
+            base=config.rotary_emb_base,
+            scale_base=config.rotary_emb_scale_base,  # If scale_base is not None, this implements XPos (Sun et al., https://arxiv.org/abs/2212.10554).
+            interleaved=config.rotary_emb_interleaved,
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Perform self-attention.
+
+        There are two attention implementations supported:
+        Flash Attention 2 and PyTorch's scaled_dot_product_attention.
+
+        Args:
+            hidden_states: (batch, seqlen, dim)
+            attn_mask: (batch, seqlen)
+
+        Returns:
+            attention: (batch, seqlen, dim)
+        """
+        batch_size, seqlen, dim = hidden_states.shape
+        qkv = self.Wqkv(hidden_states)
+
+        seqlen_offset = 0
+
+        if IMPL_USE_FLASH2:
+            # Reshape to (batch, seqlen, 3, nheads, headdim)
+            qkv = qkv.view(batch_size, seqlen, 3, self.num_attention_heads, self.attn_head_size)
+
+            # Apply RoPE
+            qkv = self.rotary_emb(qkv, seqlen_offset=seqlen_offset, max_seqlen=None)
+
+            convert_dtype = qkv.dtype not in (torch.float16, torch.bfloat16)
+            if convert_dtype:
+                # FA2 implementation only supports fp16 and bf16. If FA2 is supported,
+                # bfloat16 must be supported as of FA2 2.5.7. (Turing GPUs not supported)
+                orig_dtype = qkv.dtype
+                qkv = qkv.to(torch.bfloat16)
+
+                attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
+                attn = attn.to(orig_dtype)  # type: ignore
+            else:
+                attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
+        else:
+            qkv = qkv.view(batch_size, seqlen, 3, self.num_attention_heads, self.attn_head_size)
```

Comment:
This row (and 611) could be moved out of the if and else statement as both if and else start with the same row